File: packages/domain/src/item_type.ts

export class ItemType {
    public readonly TypeID: number;
    public readonly IsSystem: boolean;
    public Name: string;
    public Description: string;
    public RequiredAttributes: string[];

    public constructor(dto: ItemTypeDto) {
        this.TypeID = dto.type_id;
        this.IsSystem = dto.is_system;
        this.Name = dto.name;
        this.Description = dto.description;
        this.RequiredAttributes = dto.required_attributes;
    }
}

export class ItemTypeRef {
    public readonly TypeID: number;
    public readonly IsSystem: boolean;
    public Name: string;

    public constructor(dto: ItemTypeRefDto) {
        this.TypeID = dto.type_id;
        this.IsSystem = dto.is_system;
        this.Name = dto.name;
    }
}

export class ItemTypeSummary {
    public readonly TypeID: number;
    public readonly IsSystem: boolean;
    public readonly Name: string;
    public readonly RequiredAttributesCount: number;
    public readonly ItemCount: number;

    public constructor(dto: ItemTypeSummaryDto) {
        this.TypeID = dto.type_id;
        this.IsSystem = dto.is_system;
        this.Name = dto.name;
        this.RequiredAttributesCount = dto.required_attributes_count;
        this.ItemCount = dto.item_count;
    }
}

export interface ItemTypeDto {
    type_id: number;
    name: string;
    description: string;
    is_system: boolean;
    required_attributes: string[];
}

export interface ItemTypeRefDto {
    type_id: number;
    name: string;
    is_system: boolean;
}

export interface ItemTypeSummaryDto {
    type_id: number;
    name: string;
    is_system: boolean;
    required_attributes_count: number;
    item_count: number;
}

export interface AddItemTypeDto {
    name: string;
    description: string;
    required_attributes: string[];
}

export interface UpdateItemTypeDto {
    type_id: number;
    name?: string;
    description?: string;
    required_attributes?: string[];
}

File: crates/zealot-domain/src/item_type.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::{attribute::Attribute, common::id::Id};

#[derive(Debug, Clone)]
pub struct ItemType {
    pub type_id: Id,
    pub is_system: bool,
    pub name: String,
    pub description: String,
    pub required_attributes: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct ItemTypeRef {
    pub type_id: Id,
    pub is_system: bool,
    pub name: String,
}

#[derive(Debug, Clone)]
pub struct ItemTypeSummary {
    pub type_id: Id,
    pub is_system: bool,
    pub name: String,
    pub required_attributes_count: i64,
    pub item_count: i64,
}

// Send DTO

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ItemTypeDto {
    pub type_id: i64,
    pub is_system: bool,
    pub name: String,
    pub description: String,
    pub required_attributes: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ItemTypeRefDto {
    pub type_id: i64,
    pub is_system: bool,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ItemTypeSummaryDto {
    pub type_id: i64,
    pub is_system: bool,
    pub name: String,
    pub required_attributes_count: i64,
    pub item_count: i64,
}

// Receive DTOs

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AddItemTypeDto {
    pub name: String,
    pub description: String,
    pub required_attributes: Vec<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct UpdateItemTypeDto {
    pub type_id: i64,
    pub name: Option<String>,
    pub description: Option<String>,
    pub required_attributes: Option<Vec<String>>,
}

// Errors

#[derive(Debug, thiserror::Error)]
pub enum ItemTypeError {}

// Impls

impl ItemType {
    pub fn is_valid(&self, attributes: &HashMap<String, Attribute>) -> bool {
        for key in &self.required_attributes {
            if !attributes.contains_key(key) {
                return false;
            }
        }
        return true;
    }
}

impl From<&ItemType> for ItemTypeRef {
    fn from(value: &ItemType) -> Self {
        Self {
            type_id: value.type_id,
            is_system: value.is_system,
            name: value.name.clone(),
        }
    }
}

impl From<&ItemType> for ItemTypeDto {
    fn from(value: &ItemType) -> Self {
        Self {
            type_id: value.type_id.into(),
            is_system: value.is_system,
            name: value.name.clone(),
            description: value.description.clone(),
            required_attributes: value.required_attributes.clone(),
        }
    }
}

impl From<&ItemTypeRef> for ItemTypeRefDto {
    fn from(value: &ItemTypeRef) -> Self {
        Self {
            type_id: value.type_id.into(),
            is_system: value.is_system,
            name: value.name.clone(),
        }
    }
}

impl From<&ItemTypeSummary> for ItemTypeSummaryDto {
    fn from(value: &ItemTypeSummary) -> Self {
        Self {
            type_id: value.type_id.into(),
            is_system: value.is_system,
            name: value.name.clone(),
            required_attributes_count: value.required_attributes_count,
            item_count: value.item_count,
        }
    }
}
